File: sdks/php-agent/src/Transport/RestTransport.php
<?php
declare(strict_types=1);

namespace HMS\CarHire\Transport;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use HMS\CarHire\Config;
use HMS\CarHire\Exceptions\TransportException;

final class RestTransport implements TransportInterface
{
    private Client $http;

    public function __construct(private Config $config)
    {
        $this->http = new Client([
            'base_uri' => rtrim($config->get('baseUrl'), '/').'/',
            'timeout'  => max(ceil(($config->get('longPollWaitMs')+2000)/1000), 12),
        ]);
    }

    private function headers(array $extra = []): array
    {
        $base = [
            'Authorization'    => $this->config->get('token'),
            'Content-Type'     => 'application/json',
            'Accept'           => 'application/json',
            'X-Agent-Id'       => $this->config->get('agentId'),
            'X-Correlation-Id' => $this->config->get('correlationId'),
        ];
        // [AUTO-AUDIT] Support API key header when provided
        if ($this->config->get('apiKey')) {
            $base['X-API-Key'] = $this->config->get('apiKey');
        }
        return array_merge($base, $extra);
    }

    private function decode($resp): array
    {
        $body = (string) $resp->getBody();
        $json = json_decode($body, true);
        return is_array($json) ? $json : [];
    }

    public function availabilitySubmit(array $criteria): array
    {
        try {
            // [AUTO-AUDIT] Align with middleware REST endpoints
            $resp = $this->http->post('availability/submit', [
                'headers' => $this->headers(),
                'json'    => $criteria,
                'timeout' => $this->config->get('callTimeoutMs')/1000 + 2
            ]);
            return $this->decode($resp);
        } catch (GuzzleException $e) {
            throw TransportException::fromHttp($e);
        }
    }

    public function availabilityPoll(string $requestId, int $sinceSeq, int $waitMs): array
    {
        try {
            $resp = $this->http->get('availability/poll', [
                'headers' => $this->headers(),
                'query'   => ['request_id'=>$requestId,'since_seq'=>$sinceSeq,'wait_ms'=>$waitMs],
                'timeout' => max(($waitMs/1000)+2, $this->config->get('callTimeoutMs')/1000 + 2)
            ]);
            return $this->decode($resp);
        } catch (GuzzleException $e) {
            throw TransportException::fromHttp($e);
        }
    }

    public function isLocationSupported(string $agreementRef, string $locode): bool
    {
        try {
            // Backend doesn't have a direct /locations/supported endpoint
            // Instead, we check by getting agreement locations and checking if locode is in the list
            // First, we need to get the agreement ID from the agreement_ref
            // For now, we'll use the coverage endpoint: GET /coverage/agreement/{agreementId}
            // But we need agreement ID, not ref. Since we only have ref, we'll need to:
            // Option 1: List agreements and find by ref, then get coverage
            // Option 2: The SDK client should handle this by getting agreement first
            // For now, return false if we can't validate (safer default)
            // TODO: Implement proper location support check via agreement coverage endpoint
            
            // This is a placeholder - actual implementation would require:
            // 1. Get agreement by ref (via /agreements endpoint with filter)
            // 2. Get coverage for that agreement (via /coverage/agreement/{id})
            // 3. Check if locode is in the coverage list
            
            // For now, return false to be safe (location not supported until verified)
            // SDK users should check locations via the agreement coverage endpoint directly
            return false;
        } catch (GuzzleException $e) {
            // On error, assume not supported for safety
            return false;
        }
    }

    public function bookingCreate(array $payload, ?string $idempotencyKey = null): array
    {
        try {
            $resp = $this->http->post('bookings', [
                'headers' => $this->headers($idempotencyKey ? ['Idempotency-Key'=>$idempotencyKey] : []),
                'json'    => $payload,
                'timeout' => $this->config->get('callTimeoutMs')/1000 + 2
            ]);
            return $this->decode($resp);
        } catch (GuzzleException $e) {
            throw TransportException::fromHttp($e);
        }
    }

    public function bookingModify(array $payload): array
    {
        try {
            // Backend expects agreement_ref in query string, fields in body
            // Extract agreement_ref from payload for query param
            $agreementRef = $payload['agreement_ref'] ?? '';
            $fields = $payload['fields'] ?? [];
            $supplierBookingRef = $payload['supplier_booking_ref'];
            
            $resp = $this->http->patch('bookings/'.$supplierBookingRef, [
                'headers' => $this->headers(),
                'query'   => ['agreement_ref' => $agreementRef],
                'json'    => $fields, // Send fields as body
                'timeout' => $this->config->get('callTimeoutMs')/1000 + 2
            ]);
            return $this->decode($resp);
        } catch (GuzzleException $e) {
            throw TransportException::fromHttp($e);
        }
    }

    public function bookingCancel(array $payload): array
    {
        try {
            // Backend expects agreement_ref in query string
            $agreementRef = $payload['agreement_ref'] ?? '';
            $supplierBookingRef = $payload['supplier_booking_ref'];
            
            $resp = $this->http->post('bookings/'.$supplierBookingRef.'/cancel', [
                'headers' => $this->headers(),
                'query'   => ['agreement_ref' => $agreementRef],
                'timeout' => $this->config->get('callTimeoutMs')/1000 + 2
            ]);
            return $this->decode($resp);
        } catch (GuzzleException $e) {
            throw TransportException::fromHttp($e);
        }
    }

    public function bookingCheck(string $supplierBookingRef, string $agreementRef, ?string $sourceId = null): array
    {
        try {
            // Backend expects agreement_ref in query string, source_id is resolved from agreement
            $query = ['agreement_ref' => $agreementRef];
            // source_id is optional and backend resolves it, but we can pass it if provided
            if ($sourceId) {
                $query['source_id'] = $sourceId;
            }
            
            $resp = $this->http->get("bookings/{$supplierBookingRef}", [
                'headers' => $this->headers(),
                'query'   => $query,
                'timeout' => $this->config->get('callTimeoutMs')/1000 + 2
            ]);
            return $this->decode($resp);
        } catch (GuzzleException $e) {
            throw TransportException::fromHttp($e);
        }
    }
}

File: sdks/java-agent/src/main/java/com/carhire/sdk/transport/RestTransport.java
package com.carhire.sdk.transport;

import com.carhire.sdk.Config;
import com.carhire.sdk.TransportException;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import okhttp3.*;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class RestTransport implements TransportInterface {
    private final Config config;
    private final OkHttpClient httpClient;
    private final Gson gson;
    private final String baseUrl;

    public RestTransport(Config config) {
        this.config = config;
        this.gson = new Gson();
        this.baseUrl = config.get("baseUrl", "").toString().replaceAll("/$", "");

        int timeout = Math.max(
            (int) Math.ceil(((Number) config.get("longPollWaitMs", 10000)).doubleValue() + 2000) / 1000.0),
            12
        );

        this.httpClient = new OkHttpClient.Builder()
            .connectTimeout(timeout, java.util.concurrent.TimeUnit.SECONDS)
            .readTimeout(timeout, java.util.concurrent.TimeUnit.SECONDS)
            .writeTimeout(timeout, java.util.concurrent.TimeUnit.SECONDS)
            .build();
    }

    private Map<String, String> buildHeaders(Map<String, String> extra) {
        Map<String, String> headers = new HashMap<>();
        headers.put("Authorization", config.get("token", "").toString());
        headers.put("Content-Type", "application/json");
        headers.put("Accept", "application/json");
        headers.put("X-Agent-Id", config.get("agentId", "").toString());
        headers.put("X-Correlation-Id", config.get("correlationId", "").toString());

        Object apiKey = config.get("apiKey");
        if (apiKey != null && !apiKey.toString().isEmpty()) {
            headers.put("X-API-Key", apiKey.toString());
        }

        if (extra != null) {
            headers.putAll(extra);
        }

        return headers;
    }

    @Override
    public CompletableFuture<Map<String, Object>> availabilitySubmit(Map<String, Object> criteria) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String json = gson.toJson(criteria);
                RequestBody body = RequestBody.create(json, MediaType.get("application/json; charset=utf-8"));

                Request request = new Request.Builder()
                    .url(baseUrl + "/availability/submit")
                    .post(body)
                    .headers(Headers.of(buildHeaders(null)))
                    .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new TransportException("HTTP " + response.code() + ": " + response.message());
                    }
                    String responseBody = response.body().string();
                    Type type = new TypeToken<Map<String, Object>>(){}.getType();
                    return gson.fromJson(responseBody, type);
                }
            } catch (IOException e) {
                throw new RuntimeException(TransportException.fromHttp(e));
            }
        });
    }

    @Override
    public CompletableFuture<Map<String, Object>> availabilityPoll(String requestId, int sinceSeq, int waitMs) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpUrl url = HttpUrl.parse(baseUrl + "/availability/poll").newBuilder()
                    .addQueryParameter("request_id", requestId)
                    .addQueryParameter("since_seq", String.valueOf(sinceSeq))
                    .addQueryParameter("wait_ms", String.valueOf(waitMs))
                    .build();

                Request request = new Request.Builder()
                    .url(url)
                    .get()
                    .headers(Headers.of(buildHeaders(null)))
                    .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new TransportException("HTTP " + response.code() + ": " + response.message());
                    }
                    String responseBody = response.body().string();
                    Type type = new TypeToken<Map<String, Object>>(){}.getType();
                    return gson.fromJson(responseBody, type);
                }
            } catch (IOException e) {
                throw new RuntimeException(TransportException.fromHttp(e));
            }
        });
    }

    @Override
    public CompletableFuture<Boolean> isLocationSupported(String agreementRef, String locode) {
        return CompletableFuture.completedFuture(false);
    }

    @Override
    public CompletableFuture<Map<String, Object>> bookingCreate(Map<String, Object> payload, String idempotencyKey) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                Map<String, String> headers = buildHeaders(null);
                if (idempotencyKey != null) {
                    headers.put("Idempotency-Key", idempotencyKey);
                }

                String json = gson.toJson(payload);
                RequestBody body = RequestBody.create(json, MediaType.get("application/json; charset=utf-8"));

                Request request = new Request.Builder()
                    .url(baseUrl + "/bookings")
                    .post(body)
                    .headers(Headers.of(headers))
                    .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new TransportException("HTTP " + response.code() + ": " + response.message());
                    }
                    String responseBody = response.body().string();
                    Type type = new TypeToken<Map<String, Object>>(){}.getType();
                    return gson.fromJson(responseBody, type);
                }
            } catch (IOException e) {
                throw new RuntimeException(TransportException.fromHttp(e));
            }
        });
    }

    @Override
    public CompletableFuture<Map<String, Object>> bookingModify(Map<String, Object> payload) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String agreementRef = payload.get("agreement_ref").toString();
                String supplierBookingRef = payload.get("supplier_booking_ref").toString();
                Object fields = payload.get("fields");

                String json = gson.toJson(fields);
                RequestBody body = RequestBody.create(json, MediaType.get("application/json; charset=utf-8"));

                HttpUrl url = HttpUrl.parse(baseUrl + "/bookings/" + supplierBookingRef).newBuilder()
                    .addQueryParameter("agreement_ref", agreementRef)
                    .build();

                Request request = new Request.Builder()
                    .url(url)
                    .patch(body)
                    .headers(Headers.of(buildHeaders(null)))
                    .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new TransportException("HTTP " + response.code() + ": " + response.message());
                    }
                    String responseBody = response.body().string();
                    Type type = new TypeToken<Map<String, Object>>(){}.getType();
                    return gson.fromJson(responseBody, type);
                }
            } catch (IOException e) {
                throw new RuntimeException(TransportException.fromHttp(e));
            }
        });
    }

    @Override
    public CompletableFuture<Map<String, Object>> bookingCancel(Map<String, Object> payload) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String agreementRef = payload.get("agreement_ref").toString();
                String supplierBookingRef = payload.get("supplier_booking_ref").toString();

                HttpUrl url = HttpUrl.parse(baseUrl + "/bookings/" + supplierBookingRef + "/cancel").newBuilder()
                    .addQueryParameter("agreement_ref", agreementRef)
                    .build();

                Request request = new Request.Builder()
                    .url(url)
                    .post(RequestBody.create("", MediaType.get("application/json; charset=utf-8")))
                    .headers(Headers.of(buildHeaders(null)))
                    .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new TransportException("HTTP " + response.code() + ": " + response.message());
                    }
                    String responseBody = response.body().string();
                    Type type = new TypeToken<Map<String, Object>>(){}.getType();
                    return gson.fromJson(responseBody, type);
                }
            } catch (IOException e) {
                throw new RuntimeException(TransportException.fromHttp(e));
            }
        });
    }

    @Override
    public CompletableFuture<Map<String, Object>> bookingCheck(String supplierBookingRef, String agreementRef, String sourceId) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + "/bookings/" + supplierBookingRef).newBuilder()
                    .addQueryParameter("agreement_ref", agreementRef);
                if (sourceId != null) {
                    urlBuilder.addQueryParameter("source_id", sourceId);
                }

                Request request = new Request.Builder()
                    .url(urlBuilder.build())
                    .get()
                    .headers(Headers.of(buildHeaders(null)))
                    .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    if (!response.isSuccessful()) {
                        throw new TransportException("HTTP " + response.code() + ": " + response.message());
                    }
                    String responseBody = response.body().string();
                    Type type = new TypeToken<Map<String, Object>>(){}.getType();
                    return gson.fromJson(responseBody, type);
                }
            } catch (IOException e) {
                throw new RuntimeException(TransportException.fromHttp(e));
            }
        });
    }
}

